text/0000-as-keyword-consider-into-trait.md

```diff
@@ -0,0 +1,130 @@
+- the `+=`, `-=`, `*=`, `\=` operators are the result of the `AddAssign`, `SubAssign`, `MulAssign`, `DivAssign` Traits implementations.
+- the `container[index]` operator is the result of the `Index` and `IndexMut` Traits implementations.
+- and many others can be found on the [`std::ops`](https://doc.rust-lang.org/std/ops/) documentation page.
+
+The `as` operator is actually reserved to primitives only but these given primitives already implement `Into` Traits for all possible conversions:
+- the `u32` implement `Into<u64>` in the form of an [`impl From<u32> for u64`](https://doc.rust-lang.org/std/primitive.u64.html#impl-From%3Cu32%3E).
+- the `u8` implement `Into<u16>` in the form of an [`impl From<u8> for u16`](https://doc.rust-lang.org/std/primitive.u16.html#impl-From%3Cu8%3E).
+
+All of these conversions can be found in the [`libcore/num/mod.rs` file](https://github.com/rust-lang/rust/blob/d9d5c667d819ce400fc7adb09dcd6482b0aa519e/src/libcore/num/mod.rs#L3343-L3400).
+
+Some special primitives like `usize` and `isize` doesn't implement many `From` Traits because these are [the representation of the address register](https://users.rust-lang.org/t/cant-convert-usize-to-u64/6243) and it can break compilation on some architectures.
+
+Adding this feature to the compiler will probably unify the [actual conversion error detections](https://users.rust-lang.org/t/cant-convert-usize-to-u64/6243/8) of the `as` keyword and the `into` method.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The actual design is not far from what we already know of the `as` keyword.
+It doesn't add new grammar to the language, it add more freedom to the actual syntax:
+
+```rust
+struct Foo(i32);
+
+struct Bar(i32);
+
+impl Into<Bar> for Foo {
+    fn into(self) -> Bar {
+        Bar(self.0)
+    }
+}
+
+// this old syntax
+let x = Foo(42);
+let y: Bar = x.into();
+
+// is equivalent to the new one
+let x = Foo(42);
+let y = x as Bar;
+```
+
+The actual syntax **is not deprecated in any way**, it has to mimic the `Add::add` or the `Deref::deref` methods and permit the user to directly call these methods or use the operators as he likes.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `as` keyword is use to reflect the actual `Into::into` method call, if the `Into` Trait is implemented for the type to be converted, the compiler needs to emit an error informing the user that the conversion is not possible.
+
+The actual `as` keyword is reserved for primitives to convert into other primitives but it shouldn't be a breaking change to change the behavior and to work with any type implementing `Into` because the primitives already implement these Traits.
+
+The internal desing could be a simple syntax sugar to something like that:
+
+```rust
+// this actual syntax
+let x = Foo(42);
+let y = x as Bar;
+
+// is a syntax sugar for
+let x = Foo(42);
+let y = Into::<Bar>::into(x);
+```
```

Comment:
This won't work since `as` is not equivalent to a call to `.into()` - for example, `as` can convert from `f32` to `u32`, while `.into()` can't (this is a lossy conversion, so `.into()` *shouldn't* work here).

Reply:
So what is the rule of `as` conversion and `into` conversion ?

I think this is related to [this thread about lossy conversions](https://github.com/Kerollmops/rust-rfcs/blob/1285b70/text/0000-as-keyword-consider-into-trait.md) (I linked it in the Renderer).